src/pcnt.rs

```diff
@@ -0,0 +1,449 @@
+            }
+            PcntControlMode::Disable => {
+                esp_idf_sys::pcnt_channel_edge_action_t_PCNT_CHANNEL_EDGE_ACTION_DECREASE
+            }
+        }
+    }
+}
+
+bitflags! {
+    #[allow(dead_code)]
+    #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+    pub struct PcntEventType: u32 {
+        const THRES_1 = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_THRES_1;
+        const THRES_0 = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_THRES_0;
+        const L_LIM = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_L_LIM;
+        const H_LIM = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_H_LIM;
+        const ZERO = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_ZERO;
+    }
+}
+
+#[doc = " @brief Pulse Counter configuration for a single channel"]
+#[derive(Debug)]
+pub struct PcntConfig {
+    #[doc = "< Pulse input GPIO number, if you want to use GPIO16, enter pulse_gpio_num = 16, a negative value will be ignored"]
+    pub pulse_pin: Option<PcntPin>,
+    #[doc = "< Control signal input GPIO number, a negative value will be ignored"]
+    pub ctrl_pin: Option<PcntPin>,
+    #[doc = "< PCNT low control mode"]
+    pub lctrl_mode: PcntControlMode,
+    #[doc = "< PCNT high control mode"]
+    pub hctrl_mode: PcntControlMode,
+    #[doc = "< PCNT positive edge count mode"]
+    pub pos_mode: PcntCountMode,
+    #[doc = "< PCNT negative edge count mode"]
+    pub neg_mode: PcntCountMode,
+    #[doc = "< Maximum counter value"]
+    pub counter_h_lim: i16,
+    #[doc = "< Minimum counter value"]
+    pub counter_l_lim: i16,
+    #[doc = "< the PCNT channel"]
+    pub channel: PcntChannel,
+}
+
+impl Into<pcnt_config_t> for &PcntConfig {
+    fn into(self) -> pcnt_config_t {
+        pcnt_config_t {
+            pulse_gpio_num: match &self.pulse_pin {
+                Some(pin) => pin.pin_num(),
+                None => esp_idf_sys::PCNT_PIN_NOT_USED,
+            },
+            ctrl_gpio_num: match &self.ctrl_pin {
+                Some(pin) => pin.pin_num(),
+                None => esp_idf_sys::PCNT_PIN_NOT_USED,
+            },
+            lctrl_mode: self.lctrl_mode.into(),
+            hctrl_mode: self.hctrl_mode.into(),
+            pos_mode: self.pos_mode.into(),
+            neg_mode: self.neg_mode.into(),
+            counter_h_lim: self.counter_h_lim,
+            counter_l_lim: self.counter_l_lim,
+            channel: self.channel.into(),
+            ..Default::default()
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct PcntPin {
+    pin: Arc<i32>
+}
+
+impl PcntPin {
+    pub fn new<'d>(pin: impl Peripheral<P = impl IOPin> + 'd, pull_mode: Pull) -> Result<Self, EspError>
+    {
+        crate::into_ref!(pin);
+        let p = PcntPin { pin: Arc::new(pin.pin()) };
+        PinDriver::input(pin)?.into_input()?.set_pull(pull_mode)?;
+        Ok(p)
+    }
+
+    fn pin_num(&self) -> i32 {
+        *self.pin
+    }
+}
+
+
+#[derive(Debug)]
+struct PcntChannelPins {
+    pulse: Option<PcntPin>,
```

Comment:
Any way to prevent the `T: Borrow<PcntPin<'d>>` from propagating everywhere like this?
```
pub struct Pcnt<'d, T: Borrow<PcntPin<'d>>> {
    unit: pcnt_unit_t,
    channel_pins: [PcntChannelPins<'d, T>; esp_idf_sys::pcnt_channel_t_PCNT_CHANNEL_MAX as usize]
}
```

Reply:
Yes, internally I only need it for the pin number, externally I need the same one in two channels.

Reply:
This 
```
struct PcntChannelPins <'d, T: Borrow<PcntPin<'d>>>
{
    pulse: Option<T>,
    ctrl: Option<T>,
    _p: PhantomData<&'d ()>,
}

impl<'d, T: Borrow<PcntPin<'d>>> PcntChannelPins<'d, T> {
    fn get_pin_numbers(&self) -> (i32, i32) {
        (
            match &self.pulse {
                Some(pin) => pin.pin_num(),
                None => esp_idf_sys::PCNT_PIN_NOT_USED,
            },
            match &self.ctrl {
                Some(pin) => pin.pin_num(),
                None => esp_idf_sys::PCNT_PIN_NOT_USED,
            },
        )
    }
}
```
complains that "no method named `pin_num` found for reference `&T` in the current scope
method not found in `&T`" for `Some(pin) => pin.pin_num()`

Reply:
This 
```
struct PcntChannelPins <'d, T: Borrow<PcntPin<'d>>>
{
    pulse: Option<T>,
    ctrl: Option<T>,
    _p: PhantomData<&'d ()>,
}

impl<'d, T: Borrow<PcntPin<'d>>> PcntChannelPins<'d, T> {
    fn get_pin_numbers(&self) -> (i32, i32) {
        (
            match &self.pulse {
                Some(pin) => pin.pin_num(),
                None => esp_idf_sys::PCNT_PIN_NOT_USED,
            },
            match &self.ctrl {
                Some(pin) => pin.pin_num(),
                None => esp_idf_sys::PCNT_PIN_NOT_USED,
            },
        )
    }
}
```
complains that "no method named `pin_num` found for reference `&T` in the current scope
method not found in `&T`" for `Some(pin) => pin.pin_num()`

Reply:
When using Borrow you need to provide typeannotation like &PcntPin when you using it

Reply:
PcntPin now passed in as a Borrow